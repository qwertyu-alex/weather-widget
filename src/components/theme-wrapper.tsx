"use client";

import { useTheme } from "next-themes";
import { useParams, useSearchParams } from "next/navigation";

export const ThemeWrapper = (props: { children: JSX.Element }) => {
  const params = useSearchParams();
  const { setTheme } = useTheme();
  const theme = params.get("theme");

  if (theme) {
    if (theme === "dark") {
      setTheme("dark");
    }

    if (theme === "light") {
      setTheme("light");
    }
  }

  return props.children;
};
