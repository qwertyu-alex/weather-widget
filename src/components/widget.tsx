"use client";

import { zodResolver } from "@hookform/resolvers/zod";
import { useRouter, useSearchParams } from "next/navigation";
import { useCallback } from "react";
import { useForm } from "react-hook-form";
import { z } from "zod";
import { Button } from "./ui/button";
import { Form, FormControl, FormField, FormItem, FormMessage } from "./ui/form";
import { Input } from "./ui/input";

const SearchForm = (props: { placeholder: string }) => {
  const searchParams = useSearchParams();
  const router = useRouter();

  const formSchema = z.object({
    city: z.string().min(1),
  });

  const form = useForm<z.infer<typeof formSchema>>({
    resolver: zodResolver(formSchema),
    defaultValues: {
      city: "",
    },
  });

  const createQueryString = useCallback(
    (name: string, value: string) => {
      const params = new URLSearchParams(Array.from(searchParams.entries())); // -> has to use this form
      params.set(name, value);

      return params.toString();
    },
    [searchParams]
  );

  function onSubmit(values: z.infer<typeof formSchema>) {
    const queryString = createQueryString("q", values.city);
    router.push(`?${queryString}`);
  }

  return (
    <>
      <div className="script w-full p-5 flex flex-col gap-2">
        <Form {...form}>
          <fieldset>
            <form onSubmit={form.handleSubmit(onSubmit)} className="space-y-2">
              <FormField
                control={form.control}
                name="city"
                render={({ field }) => (
                  <FormItem>
                    <FormControl>
                      <Input placeholder={props.placeholder} {...field} />
                    </FormControl>
                    <FormMessage />
                  </FormItem>
                )}
              />
              <Button type="submit" className="script">
                Search for another city
              </Button>
            </form>
          </fieldset>
        </Form>
      </div>
    </>
  );
};

export const ClientWidget = (props: {
  cityName?: string;
  temperature?: number;
  humidity?: number;
  wind?: number;
}) => {
  return (
    <>
      <div className="w-full p-5 border-b bg-muted font-medium flex gap-1">
        <p>Weather in</p>
        <p>{props.cityName}</p>
      </div>

      <div className="w-full p-5 border-b flex gap-1 tabular-nums">
        <p>Temperature:</p>
        <p>{props.temperature}</p>
        <p>C</p>
      </div>

      <div className="w-full p-5 border-b flex gap-1 tabular-nums">
        <p>Humidity:</p>
        <p>{props.humidity}</p>
        <p>%</p>
      </div>

      <div className="w-full p-5 border-b flex gap-1 tabular-nums">
        <p>Wind:</p>
        <p>{props.wind}</p>
        <p>m/s</p>
      </div>

      <SearchForm placeholder={props.cityName || "City"} />
    </>
  );
};

export const ClientWidgetNotFound = (props: { q: string }) => {
  const q = props.q;

  return (
    <>
      <div className="w-full p-5 border-b bg-muted font-medium flex gap-1">
        <p>&quot;{q}&quot; not found. Please search for another city</p>
      </div>

      <SearchForm placeholder={q && q.trim().length > 0 ? q : "City"} />
    </>
  );
};

export const ClientWidgetInternalError = (props: {
  error: { status: number; message: string };
  q: string;
}) => {
  const q = props.q;

  return (
    <>
      <div className="w-full p-5 border-b bg-muted font-medium flex gap-1 text-red-400 flex-col">
        <p>
          ERROR: {props.error.status} {props.error.message}
        </p>
        <p>Please try again</p>
      </div>

      <SearchForm placeholder={q && q.trim().length > 0 ? q : "City"} />
    </>
  );
};
