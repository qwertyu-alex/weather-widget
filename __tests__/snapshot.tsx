/**
 * @jest-environment jsdom
 */
import { ClientWidget } from "@/components/widget";
import { render } from "@testing-library/react";

it("renders homepage unchanged", async () => {
  const { container } = render(<ClientWidget></ClientWidget>);
  expect(container).toMatchSnapshot();
});
