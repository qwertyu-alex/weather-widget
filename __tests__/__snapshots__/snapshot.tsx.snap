// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders homepage unchanged 1`] = `
<div>
  <div
    class="w-full p-5 border-b bg-muted font-medium flex gap-1"
  >
    <p>
      Weather in
    </p>
    <p />
  </div>
  <div
    class="w-full p-5 border-b flex gap-1 tabular-nums"
  >
    <p>
      Temperature:
    </p>
    <p />
    <p>
      C
    </p>
  </div>
  <div
    class="w-full p-5 border-b flex gap-1 tabular-nums"
  >
    <p>
      Humidity:
    </p>
    <p />
    <p>
      %
    </p>
  </div>
  <div
    class="w-full p-5 border-b flex gap-1 tabular-nums"
  >
    <p>
      Wind:
    </p>
    <p />
    <p>
      m/s
    </p>
  </div>
  <div
    class="script w-full p-5 flex flex-col gap-2"
  >
    <fieldset>
      <form
        class="space-y-2"
      >
        <div
          class="space-y-2"
        >
          <input
            aria-describedby=":r0:-form-item-description"
            aria-invalid="false"
            class="flex h-10 w-full rounded-md border border-input bg-background px-3 py-2 text-sm ring-offset-background file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50"
            id=":r0:-form-item"
            name="city"
            placeholder="City"
            value=""
          />
        </div>
        <button
          class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 bg-primary text-primary-foreground hover:bg-primary/90 h-10 px-4 py-2 script"
          type="submit"
        >
          Search for another city
        </button>
      </form>
    </fieldset>
  </div>
</div>
`;
